File: java/src/main/java/customer/PriceEngine.java
package customer;

import lombok.experimental.UtilityClass;
import lombok.val;

import static java.lang.Math.min;

@UtilityClass
public class PriceEngine {
    public static double calculateDiscount(Product... products) {
        val discount = products.length * 0.01;
        return min(discount, 0.2);
    }
}

File: scala/src/main/scala/demo/customer/PriceEngine.scala
package demo.customer

final case class Product(name: String)

object PriceEngine {
  private val maxDiscount: Double = 0.2
  private val ratio: Double = 0.01

  def calculateDiscount(products: Product*): Double = {
    val discount = products.length * ratio
    Math.min(discount, maxDiscount)
  }
}
